python/lsst/daf/butler/registry/datasets/byDimensions/_storage.py

```diff
@@ -430,3 +402,200 @@ def getDataId(self, id: int) -> DataCoordinate:
+    """Implementation of ByDimensionsDatasetRecordStorage which uses integer
+    auto-incremented column for dataset IDs.
+    """
+
+    def insert(self, run: RunRecord, dataIds: Iterable[DataCoordinate],
+               idMode: DatasetIdGenEnum = DatasetIdGenEnum.UNIQUE) -> Iterator[DatasetRef]:
+        # Docstring inherited from DatasetRecordStorage.
+        staticRow = {
+            "dataset_type_id": self._dataset_type_id,
+            self._runKeyColumn: run.key,
+        }
+        # Iterate over data IDs, transforming a possibly-single-pass iterable
+        # into a list, and remembering any governor dimension values we see.
+        governorValues = GovernorDimensionRestriction.makeEmpty(self.datasetType.dimensions.universe)
+        dataIdList = []
+        for dataId in dataIds:
+            dataIdList.append(dataId)
+            governorValues.update_extract(dataId)
+        with self._db.transaction():
+            # Insert into the static dataset table, generating autoincrement
+            # dataset_id values.
+            datasetIds = self._db.insert(self._static.dataset, *([staticRow]*len(dataIdList)),
+                                         returnIds=True)
+            assert datasetIds is not None
+            # Update the summary tables for this collection in case this is the
+            # first time this dataset type or these governor values will be
+            # inserted there.
+            self._summaries.update(run, self.datasetType, self._dataset_type_id, governorValues)
+            # Combine the generated dataset_id values and data ID fields to
+            # form rows to be inserted into the tags table.
+            protoTagsRow = {
+                "dataset_type_id": self._dataset_type_id,
+                self._collections.getCollectionForeignKeyName(): run.key,
+            }
+            tagsRows = [
+                dict(protoTagsRow, dataset_id=dataset_id, **dataId.byName())
+                for dataId, dataset_id in zip(dataIdList, datasetIds)
+            ]
+            # Insert those rows into the tags table.  This is where we'll
+            # get any unique constraint violations.
+            self._db.insert(self._tags, *tagsRows)
+        for dataId, datasetId in zip(dataIdList, datasetIds):
+            yield DatasetRef(
+                datasetType=self.datasetType,
+                dataId=dataId,
+                id=datasetId,
+                run=run.name,
+            )
+
+    def import_(self, run: RunRecord, datasets: Iterable[DatasetRef],
+                idGenerationMode: DatasetIdGenEnum = DatasetIdGenEnum.UNIQUE) -> Iterator[DatasetRef]:
+        # Docstring inherited from DatasetRecordStorage.
+
+        # We do not reuse integer IDs from other repository as it is going
+        # to clash with auto-incremented IDs from this and other repos.
+        return self.insert(
+            run,
+            (dataset.dataId for dataset in datasets),
+            idGenerationMode
+        )
```

Comment:
I'd (maybe naively) have expected this method to just `raise`, since can't actually do what the caller wants, and doing what it can do is probably so surprising that even documenting it carefully won't stop someone from using it incorrectly.  But I suppose that's just a repeat of my comment on `Registry.importDatasets`.

Reply:
For integer repo this just re-implements what it was doing before, it does not do what's expected but it sort of works in a sense that it will break when you try to import same dataId. Potentially I think we could implement more expensive re-mapping of the IDs by querying existing datasets dataIds and learning their existing their IDs instead of asking for auto-incremented ones. For now I'm happy to replace this with `raise` if this is the best way to handle it.